src/Microsoft.Framework.PropertyHelper.Internal/PropertyHelper.cs

```diff
@@ -136,34 +142,80 @@ public static PropertyHelper[] GetVisibleProperties(Type type)
+        /// </summary>
+        /// <param name="propertyInfo">propertyInfo to extract the getter for.</param>
+        /// <returns>a fast getter.</returns>
+        /// <remarks>
+        /// This method is more memory efficient than a dynamically compiled lambda, and about the
+        /// same speed.
+        /// </remarks>
+        public static Func<object, object> MakeNullSafeFastPropertyGetter(PropertyInfo propertyInfo)
+        {
+            Debug.Assert(propertyInfo != null);
 
-            Delegate callPropertyGetterDelegate;
-            if (typeInput.GetTypeInfo().IsValueType)
+            var getMethod = propertyInfo.GetMethod;
+            Debug.Assert(getMethod != null);
+            Debug.Assert(!getMethod.IsStatic);
+            Debug.Assert(getMethod.GetParameters().Length == 0);
+
+            // Instance methods in the CLR can be turned into static methods where the first parameter
+            // is open over "target". This parameter is always passed by reference, so we have a code
+            // path for value types and a code path for reference types.
+            if (getMethod.DeclaringType.GetTypeInfo().IsValueType)
             {
                 // Create a delegate (ref TDeclaringType) -> TValue
-                var delegateType = typeof(ByRefFunc<,>).MakeGenericType(typeInput, typeOutput);
-                var propertyGetterAsFunc = getMethod.CreateDelegate(delegateType);
-                var callPropertyGetterClosedGenericMethod =
-                    CallPropertyGetterByReferenceOpenGenericMethod.MakeGenericMethod(typeInput, typeOutput);
-                callPropertyGetterDelegate =
-                    callPropertyGetterClosedGenericMethod.CreateDelegate(
-                        typeof(Func<object, object>), propertyGetterAsFunc);
+                return MakeFastPropertyGetter(
+                    typeof(ByRefFunc<,>),
+                    getMethod,
+                    CallNullSafePropertyGetterByReferenceOpenGenericMethod);
             }
             else
             {
                 // Create a delegate TDeclaringType -> TValue
-                var propertyGetterAsFunc =
-                    getMethod.CreateDelegate(typeof(Func<,>).MakeGenericType(typeInput, typeOutput));
-                var callPropertyGetterClosedGenericMethod =
-                    CallPropertyGetterOpenGenericMethod.MakeGenericMethod(typeInput, typeOutput);
-                callPropertyGetterDelegate =
-                    callPropertyGetterClosedGenericMethod.CreateDelegate(
-                        typeof(Func<object, object>), propertyGetterAsFunc);
+                return MakeFastPropertyGetter(
+                    typeof(Func<,>),
+                    getMethod,
+                    CallNullSafePropertyGetterOpenGenericMethod);
             }
+        }
+
+        private static Func<object, object> MakeFastPropertyGetter(
```

Comment:
note: despite my suggestion of an additional refactoring above, nice the existing repetition is avoided using this new method